contracts/stake-cw20/src/claim.rs

```diff
@@ -0,0 +1,191 @@
+use cosmwasm_std::{Addr, BlockInfo, Deps, Order, StdResult, Storage, Uint128};
+use cw_controllers::{Claim, ClaimsResponse};
+use cw_storage_plus::{Item, Map, Prefix, PrefixBound};
+use cw_utils::{Expiration, Scheduled};
+use crate::ContractError;
+
+// TODO: revisit design (split each claim on own key?)
+pub struct Claims<'a>();
+
+pub struct NewClaims<'a> {
+    pub claims: Map<'a, (&'a Addr, Scheduled, u128), bool>,
+    pub total_claims: Map<'a, &'a Addr, Uint128>,
+    pub max_index: Option<u64>,
+    pub current_index_size: Item<'a, u64>,
+    pub total_claim_index: Map<'a, (u128, &'a Addr), bool>
+}
+
+impl<'a> NewClaims<'a> {
+    pub const fn new(storage_key: &'a str, max_index: Option<u64>) -> Self {
+        NewClaims{
+            claims: Map::new(storage_key),
+            total_claim_index: Map::new(storage_key + "_balance_index"),
+            total_claims: Map::new(storage_key + "_total_claims"),
+            current_index_size: Item::new(storage_key + "_current_index_size"),
+            max_index,
+        }
+    }
+
+    /// This creates a claim, such that the given address can claim an amount of tokens after
+    /// the release date.
+    pub fn create_claim(
+        &self,
+        storage: &mut dyn Storage,
+        addr: &Addr,
+        amount: Uint128,
+        release_at: Scheduled,
+    ) -> StdResult<()> {
+        // add a claim to this user to get their tokens after the unbonding period
+        self.claims.save(storage, (addr, release_at, amount.u128()), &true)?;
+
+        // if max index is setup, index stake
+        if let Some(max_index) = self.max_index {
+            let total_claim = self.total_claims.update(storage, addr, |old| {
+                match old {
+                    None => Ok(amount),
+                    Some(o) => Ok(amount + o)
+                }
+            })?;
+
+            // get last indexed total_claim after current claim
+            let last_claim  = self.total_claim_index
+                .prefix_range(storage, Some(PrefixBound::exclusive(total_claim)), None, Order::Ascending)
+                .map(|r| r.map(|((amount, addr), v)| (amount, addr)))
+                .take(1 as usize)
+                .collect::<StdResult<Vec<_>>>()?
+                .first();
+
+            match last_claim {
+                // if not found means first claim
+                None => {
+                    self.total_claims.save(storage, addr, &amount)?;
+                    self.total_claim_index.save(storage, (amount.u128(), addr), &true)?;
+                }
+                // if found, update total claim and index
+                Some(_) => {
+                    self.total_claims.save(storage, addr, &amount)?;
+                    self.total_claim_index.save(storage, (amount.u128(), addr), &true)?;
+
+                    let current_index_size = self.current_index_size.load(storage)?;
+                    // remove last element if current_index_size > max_size
```

Comment:
Could you help me understand what this is doing? From my read it seems like this is removing a claim if some condition is met (too many claims?). If this happens does that mean that the address entitled to the claim will not longer be able to claim?